# Week 19 Homework: Unsupervised Learning

### 1.	Create a study guide for all the supervised learning models we have gone over. Include information about how the algorithms work, how they are evaluated, and any other information you deem necessary. Use your own words. Do NOT copy mine or that of articles we have read.


In **<ins>supervised learning</ins>**, we know what the output variable should be. _Causal insights_ and _predictions_ are made through supervised ML. Supervised learning models use a target variable. This determines (or supervises) what the model optimizes for. Supervised ML models are trained and tested using input/output pairs. The data is divided into one larger group for training and one smaller group for testing.



* ***Regression*** methods are used for continuous data.

    * **Linear regression**: Finding a line that best fits the data to reveal the relationship between explanatory features (i.e., independent variables) and response variables (i.e., dependent variables)... __________
    
        * **Simple/univariate linear regression** plots the relationship between one explanatory variable and the response variable and predicts response variable data based on that explanatory variable.
        
            - *How it works:* y = mx + b --- This is the slope intercept formula. It caluates the predicted values for the response variable (y) by multiplying each data point of the explantory variable (x) by the slope of the line, then adding the y-intercept, which is the value of y at x = 0. 
            
            - *How it is evaluated:* <ins>Mean squared error (MSE)</ins>, which averages the squares of the errors to measure the distance between the regression line (i.e., predictive values) and the actual data. The errors are squared before averaging them because this normalizes the error data and resolves the issue of "negative" errors (i.e., point below the line). For instance, if a point is 5 units below the regression line, its distance from the line is 5, not -5. <ins>R-squared ($R^2$)</ins> is the coefficient of determination, and it also measures the distance between the observed data and the the regression line's predicted data. R is the coefficient of correlation, or in other words, the correlation between the explanatory variable and the response variable. It ranges between -1, which is a 100% negtive correlation and 1, which is a 100% positive correlation. $R^2$ normalizes this value, ignoring whether correlation is positive or negative and results in a value between 0 and 1. An $R^2$ score of zero means no correlation, and an $R^2$ value of 1 indicates 100% correlation.
            
        * **Multiple linear regression** plots the relationships between multiple explanatory variables and the response variable and predicts response variable data based on those explanatory variables.
        
            * *How it works:* y = mx + b... __________
            
            * *How it is evaluated:* MSE and $R^2$. In this case, it's especially important to use $R^2$ because R does not handle multiple explanatory variables well, whereas $R^2$ gives the percentage of variation in y that can explained by *all* explanatory variables combined.
            
        * **Polynomial regression** is used for data that does not closely fit a linear regression model. It curves the regression line by raising each predictive/explanatory feature or variable to an $n^{th}$ degree (not necessarily all to the same degree), which results in a curved regression line.
        
            * *How it works:* y = mx + b... __________
            
            * *How it is evaluated:* Mean squared error (MSE) and $R^2$ (R-squared)

    * **K-nearest neighbor**?

* ***Classification*** methods are used for discrete data.

    * **Logistic regression**???:
    
        * *How it works:* 
        
        * *How it is evaluated:*
        

    * **Support vector machines**:
    
        * *How it works:* 
        
        * *How it is evaluated:*
                
* A few methods can be used for discrete *or* continuous data:

    * **Decision trees**:
    
        * *How it works:* 
        
        * *How it is evaluated:*
        
        
    * **Random forest**:
    
        * *How it works:* 
        
        * *How it is evaluated:*
        
        
    * **Neural networks**:
    
        * *How it works:* 
        
        * *How it is evaluated:*
        

### 2.	Do the same as question 1 except for unsupervised learning models. Something that is necessary for many models is determining number, so specify how you determine the number of clusters for example. Apply this concept to any other algorithms it is relevant for.

In **<ins>unsupervised learning</ins>**, the goal is to predict relationships and/or outcomes. _Pattern discovery_ is done through unsupervised ML. In unsupervised ML, there is no target variable, only input features, which are used to find groups of similar observations. Unsupervised ML seeks to find patterns and relationships in unlabeled data. Working with unlabeled data helps prevent bias. Additionally, sometimes you don't even have a rough idea of what you're looking for, so labeling the data would not help you anyway, and it would be computationally "expensive" to label all of the data.

* ***Clustering*** methods are used for discrete data. Clustering groups data points together into categories...

    * **K-means clustering**: "trying to find the optimal number of clusters..."  
    
        * *How it works:* "type of partitioning clustering... partitions objects into 'k clusters'... 'k' just represents 'some number'... has some things in common w/ KNN w/r/t how it's calculated..." "the data points are assigned to the groups iteratively based on the similarity of the features provided." You generate an elbow diagram to determine the optimal number to use for *k*, which is the number of clusters. The K-means algorithm "randomly assigns a mean value for each of the k-clusters. Then, all the values are assigned to the closest mean. Finally, the means of those clusters are recalculated based on the actual values, and the process repeats until this mean is closest to the mean of the values in the cluster."
        
        * *How it is evaluated:* 
        
    * **Hierarchical clustering**:
    
        * *How it works:* "uses a tree-based structure... new clusters are formed using previously formed ones... data points are spread out,and we try to group them into a hierarchy, so it's a 'bottom-up' approach... we go from all of our values to figuring out how they should be grouped together, one cluster at a time... actually have twoo approaches: aglomerative, which is bottom-up, and divisive, which is top-down... "
        
        * *How it is evaluated:* 
        
    * **Grid-based** and **density-based clustering**: We did not go over these in detail, but Alexis said they are used for some more advanced things. In grid-based clustering, data is grouped into a grid, which can be calculated in a few different ways. It's grouped into the grid spaces according to density and some other criteria. In density-based clustering, there's somethign called "DB scan," and we determine distances somehow and use them for something... In areas that are densely populated with data, that data will be grouped together. Borders are determined by one of a variety of algorithms. The point of this is that data in regions with dense data have similarities that less populated regions do not share. This made me think of gerrymandering and how we often generalize that urban and rural residents have different interests, political priorities, etc.
    
* ***Dimensionality reduction*** methods are used for continuous data. They reduce the complexity of a model to make it easier to run and clear out unnecessary information. Additionally, it reduces the risk of overfitting... __________

    * **Linear dimensionality reduction**
    
        * **Principal component analysis** reduces the features of a data set to include only those that are most meaningfll or most likely to be meaningful. This reduces the computational "cost" of running the model... __________
        
            * *How it works:* 
            
            * *How it is evaluated:* 
            
        * **Singular value decomposition (SVD)** works well with sparse data, meaning your data set has a lot of missing or null values.
        
            * *How it works:* 
            
            * *How it is evaluated:* 
            
        * **Linear discriminant...**
        
            * *How it works:* 
            
            * *How it is evaluated:* 
            
    * **Non-linear dimensionality reduction** dimensionality reduction is used when data is not contained within a linear space. It uses a "manifold hypothesis," which states that the most meaningful information in is contained in a less complex, lower-dimensional space. A "manifold" is a curved surface, and basically this can be thought of as flattening out the curved surface.
    
        * **Isomap embedding**
        
            * *How it works:* 
            
            * *How it is evaluated:* 
            
        * ...


### 3.	Preprocess the customer data https://www.kaggle.com/carrie1/ecommerce-data. How are you handling nulls? What process(es) are you using to encode and normalize the data?

In [1]:
import numpy as np
import pandas as pd

commerce_df = pd.read_csv('data.csv', encoding= 'unicode_escape')
commerce_df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [2]:
commerce_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


#### Pre-processing step 1: Drop rows that do not have data for "CustomerID."

In [3]:
commerce_no_nan = commerce_df[commerce_df["CustomerID"].notnull()]

#### Pre-processing step 2: Drop rows that do not list "United Kingdom" in the "Country" column??? Or group by country instead?

In [4]:
#Getting number of customers from each country. Group by country, count unique customer IDs, and sort in descending order.
commerce_no_nan.groupby('Country')['CustomerID'].nunique().sort_values(ascending=False)

Country
United Kingdom          3950
Germany                   95
France                    87
Spain                     31
Belgium                   25
Switzerland               21
Portugal                  19
Italy                     15
Finland                   12
Austria                   11
Norway                    10
Channel Islands            9
Denmark                    9
Australia                  9
Netherlands                9
Cyprus                     8
Japan                      8
Sweden                     8
Poland                     6
Unspecified                4
Israel                     4
Greece                     4
USA                        4
Canada                     4
EIRE                       3
United Arab Emirates       2
Bahrain                    2
Malta                      2
Czech Republic             1
Singapore                  1
Lithuania                  1
Saudi Arabia               1
Brazil                     1
RSA                        1
Icelan

In [5]:
#Get total number of customers by counting the unique customer ID numbers.
commerce_no_nan['CustomerID'].nunique()

4372

In [6]:
#Get total number of customers in the UK.
commerce_no_nan[commerce_no_nan['Country']=='United Kingdom']['CustomerID'].nunique()

3950

In [7]:
#Get total number of customers outside of the UK.
commerce_no_nan[commerce_no_nan['Country']!='United Kingdom']['CustomerID'].nunique()

422

In [8]:
#Get percentage of customers outside of the UK.
commerce_no_nan[commerce_no_nan['Country']!='United Kingdom']['CustomerID'].nunique()/commerce_no_nan['CustomerID'].nunique()

0.09652333028362306

In [9]:
commerce_no_nan['TotalPrice'] = commerce_no_nan['Quantity']*commerce_no_nan['UnitPrice']
commerce_no_nan.head()

<ipython-input-9-dcd3502ed3e8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  commerce_no_nan['TotalPrice'] = commerce_no_nan['Quantity']*commerce_no_nan['UnitPrice']


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom       15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom       22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34

In [10]:
#Get total revenue from all customers.
commerce_no_nan['TotalPrice'].sum()

8300065.814000001

In [11]:
#Get total revenue from UK customers.
commerce_no_nan[commerce_no_nan['Country']=='United Kingdom']['TotalPrice'].sum()

6767873.393999998

In [12]:
#Get total revenue from Non-UK customers.
commerce_no_nan[commerce_no_nan['Country']!='United Kingdom']['TotalPrice'].sum()

1532192.4200000002

In [13]:
#Calculate the percentage of total revenue that is from non-UK customers.
commerce_no_nan[commerce_no_nan['Country']!='United Kingdom']['TotalPrice'].sum()/commerce_no_nan['TotalPrice'].sum()

0.18460003261848834

Non-UK customers only account for 9.65% of customers. However, they account for 18.46% of sales revenu, suggesting that, on average, the non-UK customers spend more. So it makes sense to include non-UK customers in the analysis, and the company may even want to focus on acquiring more customers from outside of the UK.

However, because many of the individual countries have such low customer counts, I think that I should bin the non-UK customers into one "Non-UK" group in the "Country" column.

In [14]:
#Get total numbr of countries.
commerce_no_nan['Country'].nunique()

37

In [15]:
#Because there are 37 countries (and thus 36 non-UK countries), 
commerce_natl_intl = pd.DataFrame(commerce_no_nan)

In [16]:
commerce_natl_intl.loc[commerce_natl_intl['Country'] != 'United Kingdom', 'Country'] = 'International'

In [17]:
commerce_natl_intl.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom       15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom       22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34

In [18]:
commerce_natl_intl.groupby('Country')['CustomerID'].nunique().sort_values(ascending=False)

Country
United Kingdom    3950
International      422
Name: CustomerID, dtype: int64

In [19]:
commerce_natl_intl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    406829 non-null  object 
 1   StockCode    406829 non-null  object 
 2   Description  406829 non-null  object 
 3   Quantity     406829 non-null  int64  
 4   InvoiceDate  406829 non-null  object 
 5   UnitPrice    406829 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      406829 non-null  object 
 8   TotalPrice   406829 non-null  float64
dtypes: float64(3), int64(1), object(5)
memory usage: 31.0+ MB


In [20]:
#Use integer encoding to change "Country" data to integers. 0 = UK, and 1 = International.

In [21]:
#Example code for something...
#newdf = df[(df.origin == "JFK")

#### Pre-processing step 3: Try vectorizing the "Description" column???

In [22]:
#First, I tried vectorizing the "Description" data, just to see how it would look.
from sklearn.feature_extraction.text import TfidfVectorizer
description_text = commerce_natl_intl["Description"]
tfidf_vec = TfidfVectorizer()
description_tfidf = tfidf_vec.fit_transform(description_text)

In [23]:
commerce_natl_intl["Description Vector"] = description_tfidf

In [24]:
commerce_natl_intl.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  \
0  12/1/2010 8:26       2.55     17850.0  United Kingdom       15.30   
1  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34   
2  12/1/2010 8:26       2.75     17850.0  United Kingdom       22.00   
3  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34   
4  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34   

                                  Description Vector  
0    (0, 915)\t0.4710680734758523\n  (0, 1059)\t0...  
1    (0, 915)\t0.4710680734758523\n  (0, 1059)\t0...  
2    (0, 915)\t0.4710680734758523\n  (0, 1059)\t0...  
3    (0, 915)\t0.4710680734758523\n  (0, 1059)\t0...  
4    (0, 915)\t0.4710680734758523\n  (0, 1059)\t0...

I don't think I can actually do anything with this, so I'll go ahead and drop the new "Description Vector" column. I just wanted to keep it in to show that I tried it.

In [25]:
commerce_natl_intl.drop(["Description Vector"], inplace=True, axis=1)

In [26]:
commerce_natl_intl.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom       15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom       22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34

#### Pre-processing step 4: Convert "InvoiceDate" to date_time format and extract month and year.

In [27]:
commerce_natl_intl["InvoiceDate"] = pd.to_datetime(commerce_natl_intl["InvoiceDate"])

In [32]:
commerce_natl_intl.describe(datetime_is_numeric=True)

Quantity                    InvoiceDate      UnitPrice  \
count  406829.000000                         406829  406829.000000   
mean       12.061303  2011-07-10 16:30:57.879140864       3.460471   
min    -80995.000000            2010-12-01 08:26:00       0.000000   
25%         2.000000            2011-04-06 15:02:00       1.250000   
50%         5.000000            2011-07-31 11:48:00       1.950000   
75%        12.000000            2011-10-20 13:06:00       3.750000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       248.693370                            NaN      69.315162   

          CustomerID     TotalPrice          Month           Year  
count  406829.000000  406829.000000  406829.000000  406829.000000  
mean    15287.690570      20.401854       7.605947    2010.934002  
min     12346.000000 -168469.600000       1.000000    2010.000000  
25%     13953.000000       4.200000       5.000000    2011.000000  
50%     15152.000000      11.100000       8.000000    2011.000000  
75%     16791.000000      19.500000      11.000000    2011.000000  
max     18287.000000  168469.600000      12.000000    2011.000000  
std      1713.600303     427.591718       3.418942       0.248279

In [28]:
commerce_natl_intl["Month"] = commerce_natl_intl["InvoiceDate"].apply(lambda row: row.month)

In [31]:
commerce_natl_intl["Year"] = commerce_natl_intl["InvoiceDate"].apply(lambda row: row.year)

In [33]:
commerce_natl_intl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    406829 non-null  object        
 1   StockCode    406829 non-null  object        
 2   Description  406829 non-null  object        
 3   Quantity     406829 non-null  int64         
 4   InvoiceDate  406829 non-null  datetime64[ns]
 5   UnitPrice    406829 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      406829 non-null  object        
 8   TotalPrice   406829 non-null  float64       
 9   Month        406829 non-null  int64         
 10  Year         406829 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(3), object(4)
memory usage: 37.2+ MB


In [34]:
commerce_natl_intl.drop(["InvoiceDate"], inplace=True, axis=1)

In [35]:
commerce_natl_intl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    406829 non-null  object 
 1   StockCode    406829 non-null  object 
 2   Description  406829 non-null  object 
 3   Quantity     406829 non-null  int64  
 4   UnitPrice    406829 non-null  float64
 5   CustomerID   406829 non-null  float64
 6   Country      406829 non-null  object 
 7   TotalPrice   406829 non-null  float64
 8   Month        406829 non-null  int64  
 9   Year         406829 non-null  int64  
dtypes: float64(3), int64(3), object(4)
memory usage: 34.1+ MB


#### Pre-processing step 5: Integer encode all columns containing string data.

According to the DataCamp course on pre-processing, it's better to use Pandas for this if you still have more pre-processing to do. However, the only preprocessing I have left is removing redundant features, and I will need to integer-encode all string data to do that, so I do not think that using LabelEncoder now will interfere with that.

In [36]:
commerce_natl_intl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    406829 non-null  object 
 1   StockCode    406829 non-null  object 
 2   Description  406829 non-null  object 
 3   Quantity     406829 non-null  int64  
 4   UnitPrice    406829 non-null  float64
 5   CustomerID   406829 non-null  float64
 6   Country      406829 non-null  object 
 7   TotalPrice   406829 non-null  float64
 8   Month        406829 non-null  int64  
 9   Year         406829 non-null  int64  
dtypes: float64(3), int64(3), object(4)
memory usage: 34.1+ MB


In [37]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
string_columns = ["InvoiceNo", "StockCode", "Description", "Country"]
commerce_natl_intl[string_columns] = commerce_natl_intl[string_columns].apply(le.fit_transform)

In [38]:
commerce_natl_intl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    406829 non-null  int64  
 1   StockCode    406829 non-null  int64  
 2   Description  406829 non-null  int64  
 3   Quantity     406829 non-null  int64  
 4   UnitPrice    406829 non-null  float64
 5   CustomerID   406829 non-null  float64
 6   Country      406829 non-null  int64  
 7   TotalPrice   406829 non-null  float64
 8   Month        406829 non-null  int64  
 9   Year         406829 non-null  int64  
dtypes: float64(3), int64(7)
memory usage: 34.1 MB


In [39]:
commerce_natl_intl.describe(datetime_is_numeric=True)

InvoiceNo      StockCode    Description       Quantity  \
count  406829.000000  406829.000000  406829.000000  406829.000000   
mean     9950.901755    1573.257501    2024.692726      12.061303   
std      5552.295845     844.054584    1089.177346     248.693370   
min         0.000000       0.000000       0.000000  -80995.000000   
25%      5060.000000     937.000000    1143.000000       2.000000   
50%     10310.000000    1522.000000    1987.000000       5.000000   
75%     14651.000000    2106.000000    2945.000000      12.000000   
max     22189.000000    3683.000000    3895.000000   80995.000000   

           UnitPrice     CustomerID        Country     TotalPrice  \
count  406829.000000  406829.000000  406829.000000  406829.000000   
mean        3.460471   15287.690570       0.889509      20.401854   
std        69.315162    1713.600303       0.313501     427.591718   
min         0.000000   12346.000000       0.000000 -168469.600000   
25%         1.250000   13953.000000       1.000000       4.200000   
50%         1.950000   15152.000000       1.000000      11.100000   
75%         3.750000   16791.000000       1.000000      19.500000   
max     38970.000000   18287.000000       1.000000  168469.600000   

               Month           Year  
count  406829.000000  406829.000000  
mean        7.605947    2010.934002  
std         3.418942       0.248279  
min         1.000000    2010.000000  
25%         5.000000    2011.000000  
50%         8.000000    2011.000000  
75%        11.000000    2011.000000  
max        12.000000    2011.000000

#### Pre-processing step 6: Delete some columns that are too strongly correlated with other columns.

In [40]:
commerce_natl_intl.corr()

InvoiceNo  StockCode  Description  Quantity  UnitPrice  \
InvoiceNo     1.000000   0.087392     0.023831 -0.009164   0.007697   
StockCode     0.087392   1.000000    -0.011504 -0.001231   0.022949   
Description   0.023831  -0.011504     1.000000  0.000174  -0.000127   
Quantity     -0.009164  -0.001231     0.000174  1.000000  -0.001238   
UnitPrice     0.007697   0.022949    -0.000127 -0.001238   1.000000   
CustomerID   -0.005835   0.003602    -0.004231 -0.003600  -0.004560   
Country       0.005831   0.009549    -0.015675 -0.011230  -0.008370   
TotalPrice   -0.010629   0.003302    -0.000019  0.916096  -0.129296   
Month         0.645556   0.064531     0.008813 -0.002477  -0.003179   
Year          0.421010   0.032383     0.022497  0.001094   0.000979   

             CustomerID   Country  TotalPrice     Month      Year  
InvoiceNo     -0.005835  0.005831   -0.010629  0.645556  0.421010  
StockCode      0.003602  0.009549    0.003302  0.064531  0.032383  
Description   -0.004231 -0.015675   -0.000019  0.008813  0.022497  
Quantity      -0.003600 -0.011230    0.916096 -0.002477  0.001094  
UnitPrice     -0.004560 -0.008370   -0.129296 -0.003179  0.000979  
CustomerID     1.000000  0.430802   -0.002274  0.030612 -0.035955  
Country        0.430802  1.000000   -0.011279  0.026441 -0.020612  
TotalPrice    -0.002274 -0.011279    1.000000 -0.001928 -0.000158  
Month          0.030612  0.026441   -0.001928  1.000000 -0.341638  
Year          -0.035955 -0.020612   -0.000158 -0.341638  1.000000

"Quantity" is very strongly correlated with "TotalPrice." This makes sense because "TotalPrice" is derived from "Quantity" and "UnitPrice." Still, I'm surprised that it was this high (0.916) and that the correlation between "TotalPrice" and "UnitPrice" was so low (-0.129). 

I think that "TotalPrice" will be more useful than "Quantity," so I will drop the "Quantity" column. 

In [42]:
commerce_natl_intl.drop(["Quantity"], inplace=True, axis=1)

In [43]:
commerce_cleaned = pd.DataFrame(commerce_natl_intl)

In [44]:
commerce_cleaned.head()

InvoiceNo  StockCode  Description  UnitPrice  CustomerID  Country  \
0          0       3249         3716       2.55     17850.0        1   
1          0       2649         3724       3.39     17850.0        1   
2          0       2855          861       2.75     17850.0        1   
3          0       2803         1813       3.39     17850.0        1   
4          0       2802         2776       3.39     17850.0        1   

   TotalPrice  Month  Year  
0       15.30     12  2010  
1       20.34     12  2010  
2       22.00     12  2010  
3       20.34     12  2010  
4       20.34     12  2010

In [45]:
commerce_cleaned.tail()

InvoiceNo  StockCode  Description  UnitPrice  CustomerID  Country  \
541904      18535       1470         2263       0.85     12680.0        0   
541905      18535       1746          695       2.10     12680.0        0   
541906      18535       2086          701       4.15     12680.0        0   
541907      18535       2087          700       4.15     12680.0        0   
541908      18535       1039          271       4.95     12680.0        0   

        TotalPrice  Month  Year  
541904       10.20     12  2011  
541905       12.60     12  2011  
541906       16.60     12  2011  
541907       16.60     12  2011  
541908       14.85     12  2011

In [52]:
commerce_cleaned.describe()

InvoiceNo      StockCode    Description      UnitPrice  \
count  406829.000000  406829.000000  406829.000000  406829.000000   
mean     9950.901755    1573.257501    2024.692726       3.460471   
std      5552.295845     844.054584    1089.177346      69.315162   
min         0.000000       0.000000       0.000000       0.000000   
25%      5060.000000     937.000000    1143.000000       1.250000   
50%     10310.000000    1522.000000    1987.000000       1.950000   
75%     14651.000000    2106.000000    2945.000000       3.750000   
max     22189.000000    3683.000000    3895.000000   38970.000000   

          CustomerID        Country     TotalPrice          Month  \
count  406829.000000  406829.000000  406829.000000  406829.000000   
mean    15287.690570       0.889509      20.401854       7.605947   
std      1713.600303       0.313501     427.591718       3.418942   
min     12346.000000       0.000000 -168469.600000       1.000000   
25%     13953.000000       1.000000       4.200000       5.000000   
50%     15152.000000       1.000000      11.100000       8.000000   
75%     16791.000000       1.000000      19.500000      11.000000   
max     18287.000000       1.000000  168469.600000      12.000000   

                Year  
count  406829.000000  
mean     2010.934002  
std         0.248279  
min      2010.000000  
25%      2011.000000  
50%      2011.000000  
75%      2011.000000  
max      2011.000000

#### Pre-processing step 7: Normalize using StandardScaler???

In [49]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler
commerce_scaled = pd.DataFrame(ss.fit_transform(commerce_cleaned), columns = commerce_cleaned.columns)

SyntaxError: cannot assign to function call (<ipython-input-49-cb13d819c4cf>, line 3)

### 4.	Perform k-means on this dataset for customer segments. Customer segments help determine what types of people buy your product, which allows you to target more people like your usual customers. Should you look at all the data, or which subset of data should you use? What is the ideal number of clusters? Which approach did you use to find the ideal number of clusters and why?

In [ ]:
# Use Alexis's code from class!
from sklearn.cluster import KMeans
X = crime_df.drop('State',axis=1)
y = crime_df['State']

plt.scatter(X['Murder'],X['UrbanPop'])

for i,txt in enumerate(y):
    plt.annotate(txt, (X['Murder'][i],X['UrbanPop'][i]))
plt.xlabel('UrbanPop')
plt.ylabel('Murder')
plt.show()

In [ ]:
inertias = []
distortions = []
for k in range(1,10):
    #build and fit the model
    model = KMeans(n_clusters=k).fit(X)
    
    #average of squared distance from cluster center to the data points
    distortion = sum(np.min(cdist(X, model.cluster_centers_,
                                  'euclidean'),axis =1))/X.shape[0]
    
    distortions.append(distortion)
    
    inertia = model.inertia_
    inertias.append(inertia)

In [ ]:
#Code from study group:
from sklearn.cluster import KMeans
wcss = []
for k in range(1,11):
    kmeans = KMeans(n_clusters=k, init="k-means++")
    kmeans.fit(grouped_customerid.iloc[:,1:])
    wcss.append(kmeans.inertia_)
plt.figure(figsize=(12,6))    
plt.grid()
plt.plot(range(1,11),wcss, linewidth=2, color="red", marker ="8")
plt.xlabel("K Value")
plt.xticks(np.arange(1,11,1))
plt.ylabel("WCSS")
plt.show()

In [ ]:
#Scatter plot with clusters = 3, based on elbow graph...?

### 5.	Perform hierarchical clustering for customer segments. What is the ideal number of clusters? Which approach did you use to find the ideal number of clusters and why?

### 6.	Compare the results of 4 and 5. Which approach do you think is best? Why? 